packages/seacas/libraries/ioss/src/Ioss_DynamicTopology.h

```diff
@@ -0,0 +1,156 @@
+#pragma once
+
+#include "Ioss_DatabaseIO.h"      // for DatabaseIO
+#include "Ioss_DBUsage.h"
+#include "Ioss_EntityType.h"      // for EntityType, etc
+#include "Ioss_Field.h"
+#include "Ioss_GroupingEntity.h" // for GroupingEntity
+#include "Ioss_MeshType.h"
+#include "Ioss_ParallelUtils.h"   // for ParallelUtils
+#include "Ioss_Property.h" // for Property
+#include "Ioss_State.h"    // for State
+#include <assert.h>
+#include <cstddef> // for size_t, nullptr
+#include <cstdint> // for int64_t
+
+#include "Ioss_CodeTypes.h"
+#include "Ioss_Utils.h"
+#include "Ioss_VariableType.h"
+#include "ioss_export.h"
+#if !defined BUILT_IN_SIERRA
+#include <fmt/ostream.h>
+#endif
+#include <functional> // for less
+#include <iosfwd>     // for ostream
+#include <map>        // for map, map<>::value_compare
+#include <sstream>
+#include <string>  // for string, operator<
+#include <utility> // for pair
+#include <vector>  // for vector
+
+namespace Ioss {
+  class Region;
+
+  /*! The TopologyModified enumeration is used as an argument to the
+   *  topology_modified() functions in io to
+   *  specify the type of topology modification that has ocurred.  The
+   *  calls to topology_modified() are cumulative between
+   *  output steps, so a call with TOPOLOGY_REORDER followed by a call
+   *  with TOPOLOGY_SHUFFLE will do the right thing.  Typical examples
+   *  of when these would be used are:
+   *  - TOPOLOGY_SAME: No change, but easier to call function than not.
+   *  - TOPOLOGY_REORDER: Element Death which reorders the Registrars
+   *  - TOPOLOGY_SHUFFLE: Load Balancing
+   *  - TOPOLOGY_HADAPT: H-Adaptivity
+   *  - TOPOLOGY_GHOST: Ghost nodes/edges/faces/elements created/destroyed
+   *  - TOPOLOGY_GEOMETRY: Model data is modified, overlap removal.
+   *  - TOPOLOGY_CREATEDELETE: Surface erosion, particle creation
+   *  - TOPOLOGY_UNKNOWN: Something else, catchall option.
+   */
+  enum TopologyModified {
+    TOPOLOGY_SAME       = (       0), //!< No change, also used for initialization
+    TOPOLOGY_REORDER    = (1U <<  0), //!< Data structures reordered on processor, no change between procs.
+    TOPOLOGY_SHUFFLE    = (1U <<  1), //!< Globally the same, data moved among processors.
+    TOPOLOGY_HADAPT     = (1U <<  2), //!< Elements split/combined; not moved cross-proc
+    TOPOLOGY_GHOST      = (1U <<  3), //!< Ghost entities created/destroyed
+    TOPOLOGY_GEOMETRY   = (1U <<  4), //!< Geometry (mesh coordinates) modified. Restart needs to know this.
+    TOPOLOGY_CREATEFACE = (1U <<  5), //!< Face/Edge are created/deleted.
+    TOPOLOGY_CREATEELEM = (1U <<  6), //!< Elements are created/deleted.
+    TOPOLOGY_CREATENODE = (1U <<  7), //!< Nodes are created/deleted.
+    TOPOLOGY_UNKNOWN    = (1U <<  8), //!< Unknown change, recreate from scratch.
+    TOPOLOGY_AUXILIARY  = (1U <<  9), //!< An AUXILIARY relation was created/modified.
+    TOPOLOGY_CONSTRAINT = (1U << 10)  //!< Contact constraints
+  };
+
+  enum class FileControlOption {
```

Comment:
Not sure what each of these mean...  Some doxygen documentation would be good.

Reply:
These were lifted directly from Sierra and after talking to Alan, some of these will become redundant after the work he is doing to get STK and Fmwk to automatically detect topology changes in mesh-mod. I can pare this list down but it's mostly in anticipation of how the apps are going to use this ...

Reply:
The comment should have been referring to the `FileControlOption` which I think is new here.

Reply:
> The comment should have been referring to the `FileControlOption` which I think is new here.

Oh, ok ... I'll add that as I go along